umbral/curvebn.py

```diff
@@ -85,19 +85,34 @@ def hash_to_bn(cls, *crypto_items, params=None):
                     raise TypeError("{} is not acceptable type, received {}".format(item, type(item)))
                 item_bytes = item
             blake2b.update(item_bytes)
+        
+        blake2b.update(_iteration.to_bytes(4, byteorder='big'))
 
-        i = 0
-        h = 0
-        while h < params.CURVE_MINVAL_HASH_512:
-            blake2b_i = blake2b.copy()
-            blake2b_i.update(i.to_bytes(params.CURVE_KEY_SIZE_BYTES, 'big'))
-            hash_digest = blake2b_i.finalize()
-            h = int.from_bytes(hash_digest, byteorder='big', signed=False)
-            i += 1
-
-        hash_bn = h % int(params.order)
-        res = cls.from_int(hash_bn, params.curve)
-        return res
+        hash_digest = blake2b.finalize()
+        hash_digest = int.from_bytes(hash_digest, byteorder='big', signed=False)
+
+        curve_nid = backend._elliptic_curve_to_nid(params.curve)
+        order = openssl._get_ec_order_by_curve_nid(curve_nid)
+
+        try:
+            hash_digest = openssl._int_to_bn(hash_digest)
+            bn_from_hash = openssl._get_new_BN()
+            with backend._tmp_bn_ctx() as bn_ctx:
+                res = backend._lib.BN_mod(
+                    bn_from_hash, hash_digest, order, bn_ctx
+                )
+                backend.openssl_assert(res == 1)
+
+            group = openssl._get_ec_group_by_curve_nid(curve_nid)
+            bn_from_hash = cls(bn_from_hash, curve_nid, group, order)
+
+        except ValueError:
+            # This case is only reached when the result is 0, 
+            # which happens with prob. 1/order of the curve
```

Comment:
`1/order` (rather than `1 / 2 ** order`) probability could be non-negligible. Whether it is or not - I'll leave up to @cygnusv and @tuxxy to decide

Reply:
The order of the curve is already a number in the order of `2**n`. For example, for `secp256k1`, the order of the curve is approximately `2**255.99999`

Reply:
Ah. All right, in this case `1 / order` is totally fine, of course